addons/sms/models/sms_sms.py

```diff
@@ -96,20 +97,25 @@ def action_set_error(self, failure_type):
     def action_set_outgoing(self):
         self._update_sms_state_and_trackers('outgoing', failure_type=False)
 
-    def send(self, unlink_failed=False, unlink_sent=True, auto_commit=False, raise_exception=False):
+    def send(self, unlink_failed=False, unlink_sent=True, auto_commit=None, raise_exception=False):
         """ Main API method to send SMS.
 
+        This contacts an external server. If the transaction fails, it may be
+        retried which can result in sending multiple SMS messages!
+
           :param unlink_failed: unlink failed SMS after IAP feedback;
           :param unlink_sent: unlink sent SMS after IAP feedback;
-          :param auto_commit: commit after each batch of SMS;
+          :param auto_commit: deprecated, use CRON if needed to commit
           :param raise_exception: raise if there is an issue contacting IAP;
         """
-        self = self.filtered(lambda sms: sms.state == 'outgoing' and not sms.to_delete)
-        for batch_ids in self._split_batch():
-            self.browse(batch_ids)._send(unlink_failed=unlink_failed, unlink_sent=unlink_sent, raise_exception=raise_exception)
-            # auto-commit if asked except in testing mode
-            if auto_commit is True and not modules.module.current_test:
-                self._cr.commit()
+
+        if auto_commit is not None:
+            warnings.warn("Since 19.0, parameter auto_commit is deprecated", DeprecationWarning)
+        domain = [('state', '=', 'outgoing'), ('to_delete', '!=', True)]
+        records = self.try_lock_for_update().filtered_domain(domain)
+        for batch_ids in tools.split_every(self._get_send_batch_size(), records.ids):
+            records = self.browse(batch_ids)
+            records._send(unlink_failed=unlink_failed, unlink_sent=unlink_sent, raise_exception=raise_exception)
```

Comment:
Shoudln't we void memory using 'invalidate_all' ? Otherwise the loop-based will anyway fill memory if we don't void it, nullifying the whole split purpose. We had the issue with mailing composer notably.

Reply:
As I understood it, the split is to avoid overloading the remote server. The cron does separate executions for each batch so it does not have that issue.
Are you sure the invalidate is needed?

Reply:
Aren't browsed values kept in cache hence leading to memory errors ? That's what we had in mailings.

Reply:
Yes, but sms messages should be smaller than e-mails. I could `invalidate_recordset` as `invalidate_all` will also invalidate all other models and this function is often called in a business process which will already take more space in memory than all related SMS messages.